VALIDATING.md

```diff
@@ -0,0 +1,218 @@
+    # This only differs from `subResource` if the webhook specified `matchPolicy: Equivalent` and the
+    # original request to the API server was converted to a version the webhook registered for.
+    "requestSubResource": "scale",
+
+    # Name of the resource being modified
+    "name": "my-deployment",
+    # Namespace of the resource being modified, if the resource is namespaced (or is a Namespace object)
+    "namespace": "my-namespace",
+
+    # operation can be CREATE, UPDATE, DELETE, or CONNECT
+    "operation": "UPDATE",
+
+    "userInfo": {
+      # Username of the authenticated user making the request to the API server
+      "username": "admin",
+      # UID of the authenticated user making the request to the API server
+      "uid": "014fbff9a07c",
+      # Group memberships of the authenticated user making the request to the API server
+      "groups": ["system:authenticated","my-admin-group"],
+      # Arbitrary extra info associated with the user making the request to the API server.
+      # This is populated by the API server authentication layer and should be included
+      # if any SubjectAccessReview checks are performed by the webhook.
+      "extra": {
+        "some-key":["some-value1", "some-value2"]
+      }
+    },
+
+    # object is the new object being admitted.
+    # It is null for DELETE operations.
+    "object": {"apiVersion":"autoscaling/v1","kind":"Scale",...},
+    # oldObject is the existing object.
+    # It is null for CREATE and CONNECT operations.
+    "oldObject": {"apiVersion":"autoscaling/v1","kind":"Scale",...},
+    # options contains the options for the operation being admitted, like meta.k8s.io/v1 CreateOptions, UpdateOptions, or DeleteOptions.
+    # It is null for CONNECT operations.
+    "options": {"apiVersion":"meta.k8s.io/v1","kind":"UpdateOptions",...},
+
+    # dryRun indicates the API request is running in dry run mode and will not be persisted.
+    # Webhooks with side effects should avoid actuating those side effects when dryRun is true.
+    # See http://k8s.io/docs/reference/using-api/api-concepts/#make-a-dry-run-request for more details.
+    "dryRun": false
+  }
+}
+}]
+```
+
+Response example:
+```
+cat <<EOF >> $VALIDATING_RESPONSE_PATH
+{"allowed": true}
+EOF
+```
+
+Customize error message:
+```
+cat <<EOF >> $VALIDATING_RESPONSE_PATH
+{"allowed": false,
+"status": {
+      "code": 403,
+      "message": "You cannot do this because it is Tuesday and your name starts with A"
+    }
+ }
+EOF
+```
+
+
+## HTTP server and Kubernetes configuration
+
+Shell-operator should create an HTTP endpoint with TLS support and register endpoints in the ValidatingWebhookConfiguration resource.
+
+There should be a Service for shell-operator (see [Availability](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#availability)).
+
+New options:
+
+```
+  --validating-webhook-server-cert="/validating-certs/server.crt"
```

Comment:
And certificate creation?

Reply:
This automation requires an [ability to approve CSRs]((https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/#a-word-of-warning-on-the-approval-permission)) for shell-operator ServiceAccount. Is it ok even for a "magic quick start"?

Reply:
I second zuzzas's opinion. We need to automate this.
Ability to approve CSRs is not actually required. It is possible to generate CA first, then generate the certificate. The generated CA goes to the `clientConfig.caBundle` field of the ValidatingWebhookConfiguration.

We can leave these flags be, but an automatic certificate generation is highly required.

Reply:
We decided not to do it because it is over complicated, concealed from a user, and causes conflicts for HA installations.